src/App.tsx

```diff
@@ -1,222 +1,206 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React, { useContext, useEffect, useRef } from 'react';
+import React, {
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import {
+  changeTodoStatus,
+  createTodo,
+  deleteTodo,
+  getTodos,
+} from './api/todos';
 import { AuthContext } from './components/Auth/AuthContext';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Footer } from './components/Footer/Footer';
+import { NewTodo } from './components/NewTodo/NewTodo';
+import { TodoList } from './components/Todo';
+import { Todo } from './types/Todo';
+import { TodosFilter } from './types/TodosFilter_Enum';
 
 export const App: React.FC = () => {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
 
+  const [update, setUpdate] = useState(0);
+
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterTodos, setFilterTodos] = useState<TodosFilter>(TodosFilter.All);
+
+  const [alertText, setAlertText] = useState('');
+  const [isAlertVisible, setAlertVisible] = useState(false);
+  const [alertTimerId, setAlertTimerId] = useState<NodeJS.Timeout | null>(null);
+
+  const getVisibleTodos = () => {
+    switch (filterTodos) {
+      case TodosFilter.Active:
+        return todos.filter((todo) => !todo.completed);
+      case TodosFilter.Completed:
+        return todos.filter((todo) => todo.completed);
+      default:
+        return todos;
+    }
+  };
+
+  const visibleTodos = useMemo(getVisibleTodos, [todos, filterTodos]);
+
+  const clearAlert = () => {
+    if (alertTimerId !== null) {
+      clearTimeout(alertTimerId);
+      setAlertTimerId(null);
+    }
+
+    setAlertVisible(false);
+  };
+
+  const showAlert = (error: string) => {
+    clearAlert();
+    setAlertText(error);
+    setAlertVisible(true);
+    setAlertTimerId(setTimeout(() => {
+      setAlertVisible(false);
+    }, 3000));
+  };
+
+  const handleClearAlert = () => clearAlert();
+
+  const handleDeleteError = () => showAlert('Unable to add Todo');
+  const handleUpdateError = () => showAlert('Unable to update Todo');
+
+  const forceUpdate = () => {
+    setUpdate(prevValue => prevValue + 1);
+  };
+
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
   }, []);
 
+  useEffect(() => {
+    if (user) {
+      getTodos(user.id)
+        .then(todosFromServer => {
```

Comment:
Let’s use `async await` all over the code. It gives you a simpler way to work with asynchronous promise-based code.